src/Types/EnumType.php

```diff
@@ -0,0 +1,161 @@
+use Doctrine\DBAL\Platforms\SQLServerPlatform;
+use Doctrine\DBAL\Types\Exception\InvalidFormat;
+use Doctrine\DBAL\Types\Exception\InvalidType;
+use Doctrine\DBAL\Types\Exception\ValueNotConvertible;
+use ReflectionClass;
+use Stringable;
+use Throwable;
+use UnitEnum;
+
+use function array_map;
+use function class_exists;
+use function enum_exists;
+use function implode;
+use function in_array;
+use function is_object;
+use function is_string;
+use function sprintf;
+
+final class EnumType extends Type
+{
+    public string $name = 'enum';
+
+    public ?string $enumClassname = null;
+
+    /** @var array<int, string> */
+    public array $members = [];
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getMappedDatabaseTypes(AbstractPlatform $platform): array
+    {
+        return [$this->name];
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getSQLDeclaration(array $column, AbstractPlatform $platform): string
+    {
+        $values = implode(
+            ', ',
+            array_map(
+                static fn (string $value) => sprintf('\'%s\'', $value),
+                $column['members'] ?: $column['type']->members,
+            ),
+        );
+
+        return match (true) {
+            $platform instanceof SQLitePlatform => sprintf('TEXT CHECK(%s IN (%s))', $column['name'], $values),
+            $platform instanceof PostgreSQLPlatform, $platform instanceof SQLServerPlatform => sprintf('VARCHAR(255) CHECK(%s IN (%s))', $column['name'], $values),
+            default => sprintf('ENUM(%s)', $values),
+        };
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function convertToDatabaseValue($value, AbstractPlatform $platform): ?string
+    {
+        if ($value === null) {
+            return null;
+        }
+
+        if ($this->enumClassname === null) {
+            if (! is_string($value)) {
+                throw InvalidType::new($value, $this->name, ['null', 'string']);
+            }
+
+            if (! in_array($value, $this->members)) {
+                throw InvalidType::new($value, $this->name, ['null', 'string']);
+            }
+
+            return $value;
+        }
+
+        if (enum_exists($this->enumClassname)) {
+            if (! $value instanceof UnitEnum) {
+                throw InvalidType::new($value, $this->name, ['null', $this->enumClassname]);
+            }
+
+            if ($value instanceof BackedEnum) {
+                return (string) $value->value;
+            }
+
+            return $value->name;
+        }
+
+        if (is_object($value)) {
+            if ($value::class !== $this->enumClassname) {
+                throw InvalidType::new($value, $this->name, ['null', $this->enumClassname]);
+            }
+
+            if (! $value instanceof Stringable) {
+                throw new ConversionException(sprintf('Class %s must implement Stringable interface', $this->enumClassname));
+            }
+        }
+
+        return (string) $value;
```

Comment:
Again, no guesswork please. This type is responsible for writing enums into the database. If the passed value is not an enum, this class shouldn't guess what to do with it.